### DISCUSSION TITLE
[v4] Beta 5 regression with Vite import aliases

### DISCUSSION THREAD

**Discussion by jacksleight** [tailwindlabs/tailwindcss#15309] [Help]
Since https://github.com/tailwindlabs/tailwindcss/issues/15159 was fixed in beta 3 imports with aliases have worked correctly.

However it looks like something in beta 5 has broken them.

I have an alias and imports like this:

```js
config.resolve = {
    alias: {
        '@package': '/vendor/package/frontend',
    },
};
```

```css
@import '@package/css/bootstrap.css';
```

Before beta 5 they worked. Since beta 5 I now get an error:

```
[@tailwindcss/vite:generate:build] ENOENT: no such file or directory, open '/vendor/package/frontend/css/bootstrap.css'
```

The same import alias in JS files still work as expected, eg:

```js
import '@package/js/bootstrap.js';
```

**Comment by philipp-spiess** [comment]
Hey! What file was this loading before? If `@package` resolves to `/vendor/package/frontend`, doesn't `/vendor/package/frontend/css/bootstrap.css` seem correct or what am I missing here? If you want the root to be inside your project directly, I _think_ you might want to do something like this instead:

```ts
import tailwindcss from '@tailwindcss/vite'
import { defineConfig } from 'vite'
import { fileURLToPath } from 'node:url'

export default defineConfig({
  plugins: [tailwindcss()],
  resolve: {
    alias: {
      '#css-alias': fileURLToPath(new URL('./src/alias.css', import.meta.url)),
    },
  },
})
```

The Vite docs mention this too: https://vite.dev/config/shared-options#resolve-alias

> When aliasing to file system paths, always use absolute paths. Relative alias values will be used as-is and will not be resolved into file system paths.

**Reply by jacksleight** [reply]
Hmmm, I was under the impression the alias paths are absolute *to the project root*, not the whole system. So `'@package': '/vendor/package/frontend',` should resolve to `/[project_root]/vendor/package/frontend`, is that not right?

That is what's happening when I import in JS, this `import '@package/js/bootstrap.js';` imports `/[project_root]/vendor/package/frontend/js/bootstrap.js`, but in CSS the equivalent import is erroring, even though `/[project_root]/vendor/package/frontend/css/bootstrap.css` does exist.

Like I say `@import '@package/css/bootstrap.css';`was working from beta 3 to beta 5.

**Reply by philipp-spiess** [reply]
We did change something in how we use the Vite resolver in the last Beta that might have caused this change: https://github.com/tailwindlabs/tailwindcss/pull/15279 however, that fix unfortunately is going to be necessary in order to support SSR in some environments, so it's not something we can easily revert. 

Are you able to find any documentation around that absolute path behavior? I do find it a bit unexpected since that means it's not really possible to load paths outside of the current project directory? 🤔

**Reply by jacksleight** [reply]
That's a good question, absolute but from the project root is just what I've always done and has always seemed to work, but perhaps that's not actually the official way to do it? Curious why the JS import works though, perhaps it checks both root from system _and_ root from project? Will see if I can find any clarification on it.

**Reply by philipp-spiess** [reply]
@jacksleight Thanks! From our perspective, all we do is call the a Vite API that will take care of the resolution for us: https://github.com/tailwindlabs/tailwindcss/blob/85da88f85138b6e91478b56e7b27e598b265bd78/packages/%40tailwindcss-vite/src/index.ts#L38-L45

**Reply by jacksleight** [reply]
I've not been able to find a definitive answer sadly, the actual module resolution is handled by Rollup under the hood so it must be something to do with how that works.

What I can confirm is that this JS import: `import '@package/js/bootstrap.js'` works whether the alias is set to this:

```js
 '@package': '/vendor/package/frontend',
```

OR this:

```js
 '@package': '/Users/jack/project/vendor/package/frontend',
```

Which would suggest that Vite/Rollup look at both paths to resolve the import. I think 😅 .

Why the same thing doesn't happen for the CSS import (since you're just calling the Vite API) is a bit of a mystery though. 😕

I can just do this as a workaround, so not a big deal:

```js
 '@package': __dirname + '/vendor/package/frontend',
```
